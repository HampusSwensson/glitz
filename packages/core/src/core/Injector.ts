import { FontFace, Properties, PropertiesList } from '@glitz/type';
import { formatRule } from '../utils/format';
import { parseDeclarationBlock } from '../utils/parse';

export const ANIMATION_NAME = 'animationName';
export const FONT_FAMILY = 'fontFamily';

export default class Injector {
  public injectClassName: (declarations: Properties, pseudo?: string) => string;
  public injectKeyframes: (declarationList: PropertiesList) => string;
  public injectFontFace: (declarations: FontFace) => string;
  constructor(
    plainDictionary: { [block: string]: string },
    pseudoDictionary: { [pseudo: string]: { [block: string]: string } },
    keyframesDictionary: { [blockList: string]: string },
    fontFaceDictionary: { [block: string]: string },
    fontFaceOriginalDictionary: { [block: string]: string },
    incrementClassHash: () => string,
    incrementKeyframesHash: () => string,
    incrementFontFaceHash: () => string,
    injectNewClassRule?: (className: string, block: string, pseudo?: string) => void,
    injectNewKeyframesRule?: (name: string, blockList: string) => void,
    injectNewFontFaceRule?: (block: string) => void,
  ) {
    this.injectClassName = (declarations, pseudo) => {
      const block = parseDeclarationBlock(declarations);
      if (block) {
        const dictionary = pseudo ? (pseudoDictionary[pseudo] = pseudoDictionary[pseudo] || {}) : plainDictionary;
        const existingClassName = dictionary[block];
        if (existingClassName) {
          return existingClassName;
        } else {
          const className = incrementClassHash();
          dictionary[block] = className;
          if (injectNewClassRule) {
            injectNewClassRule(className, block, pseudo);
          }
          return className;
        }
      }
      return '';
    };

    this.injectKeyframes = declarationList => {
      let blockList = '';
      for (const identifier in declarationList) {
        const keyframeBlock = parseDeclarationBlock(declarationList[identifier]);
        blockList += formatRule(identifier, keyframeBlock);
      }
      if (blockList) {
        const existingName = keyframesDictionary[blockList];
        if (existingName) {
          return existingName;
        } else {
          const name = incrementKeyframesHash();
          keyframesDictionary[blockList] = name;
          if (injectNewKeyframesRule) {
            injectNewKeyframesRule(name, blockList);
          }
          return name;
        }
      }
      return '';
    };

    this.injectFontFace = originalDeclarations => {
      if (process.env.NODE_ENV !== 'production') {
        if ('fontFamily' in originalDeclarations) {
          console.warn('The CSS property `font-family` font face will be ignored in %O', originalDeclarations);
        }
      }
      delete (originalDeclarations as any)[FONT_FAMILY];
      const originalBlock = parseDeclarationBlock(originalDeclarations);
      if (originalBlock) {
        const existingClassName = fontFaceOriginalDictionary[originalBlock];
        if (existingClassName) {
          return existingClassName;
        } else {
          const name = incrementFontFaceHash();
          const declarations = { ...originalDeclarations, [FONT_FAMILY]: name };
          const block = parseDeclarationBlock(declarations);
          fontFaceDictionary[block] = fontFaceOriginalDictionary[originalBlock] = name;
          if (injectNewFontFaceRule) {
            injectNewFontFaceRule(block);
          }
          return name;
        }
      }
      return '';
    };
  }
}
