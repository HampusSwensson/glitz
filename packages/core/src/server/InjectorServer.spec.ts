import Injector from './InjectorServer';

describe('server', () => {
  it('injects plain rule', () => {
    const injector = new Injector();

    expect(injector.injectClassRule({ color: 'red' })).toBe('a');
    expect(injector.injectClassRule({ color: 'green', backgroundColor: 'black' })).toBe('b');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('injects pseudo rule', () => {
    const injector = new Injector();

    expect(injector.injectClassRule({ color: 'red' }, ':hover')).toBe('a');
    expect(injector.injectClassRule({ color: 'green', backgroundColor: 'black' }, ':hover')).toBe('b');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('injects keyframes rule', () => {
    const injector = new Injector();

    expect(injector.injectKeyframesRule({ from: { color: 'red' }, to: { color: 'green' } })).toBe('a');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('injects fallback rule', () => {
    const injector = new Injector();

    expect(injector.injectClassRule({ color: ['red', 'green'] })).toBe('a');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('reuses plain rule', () => {
    const injector = new Injector();

    expect(injector.injectClassRule({ color: 'red' })).toBe('a');
    expect(injector.injectClassRule({ color: 'red' })).toBe('a');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('reuses pseudo rule', () => {
    const injector = new Injector();

    expect(injector.injectClassRule({ color: 'red' }, ':hover')).toBe('a');
    expect(injector.injectClassRule({ color: 'red' }, ':hover')).toBe('a');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('reuses keyframes rule', () => {
    const injector = new Injector();

    expect(injector.injectKeyframesRule({ from: { color: 'red' }, to: { color: 'green' } })).toBe('a');
    expect(injector.injectKeyframesRule({ from: { color: 'red' }, to: { color: 'green' } })).toBe('a');
    expect(injector.getStyle()).toMatchSnapshot();
  });
  it('reuses fallback rule', () => {
    const injector = new Injector();

    expect(injector.injectClassRule({ color: ['red', 'green'] })).toBe('a');
    expect(injector.injectClassRule({ color: ['red', 'green'] })).toBe('a');
    expect(injector.getStyle()).toMatchSnapshot();
  });
});
