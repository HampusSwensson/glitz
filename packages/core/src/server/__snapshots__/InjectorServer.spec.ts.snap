// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server injects fallback rule 1`] = `".a{color:red;color:green}"`;

exports[`server injects font face rule 1`] = `"@font-face {font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:a}"`;

exports[`server injects keyframes rule 1`] = `"@keyframes a{from{color:red}to{color:green}}"`;

exports[`server injects plain rule 1`] = `".a{color:red}.b{color:green;background-color:black}"`;

exports[`server injects pseudo rule 1`] = `".a:hover{color:red}.b:hover{color:green;background-color:black}"`;

exports[`server reuses fallback rule 1`] = `".a{color:red;color:green}"`;

exports[`server reuses keyframes rule 1`] = `"@keyframes a{from{color:red}to{color:green}}"`;

exports[`server reuses plain rule 1`] = `".a{color:red}"`;

exports[`server reuses pseudo rule 1`] = `".a:hover{color:red}"`;
